File: 유틸리티/데이터/Value/Modifer/NumericIModifier.cs
using System;

using UnityEngine;

namespace inonego.Modifier
{
    // ============================================================
    /// <summary>
    /// 수치 연산 타입을 정의합니다.
    /// </summary>
    // ============================================================
    public enum NumericIOperation
    {
        SET, ADD, SUB, MUL, DIV
    }

    // ============================================================
    /// <summary>
    /// Int 타입의 값에 수치 연산을 적용하는 수정자입니다.
    /// </summary>
    // ============================================================
    [Serializable]
    public class NumericIModifier : IModifier<int>
    {
        [SerializeField]
        private NumericIOperation operation;
        public NumericIOperation Operation => operation;

        [SerializeField]
        private int value;
        public int Value => value;

        private NumericIModifier() {}

        public NumericIModifier(NumericIOperation operation, int value)
        {
            this.operation = operation;
            this.value = value;
        }

        public int Modify(int value)
        {
            return operation switch
            {
                NumericIOperation.SET => this.value,
                NumericIOperation.ADD => value + this.value,
                NumericIOperation.SUB => value - this.value,
                NumericIOperation.MUL => value * this.value,
                NumericIOperation.DIV => value / this.value,
                _ => value
            };
        }

        public IModifier<int> @new() => new NumericIModifier();

        public IModifier<int> Clone()
        {
            var cloned = @new();
            cloned.CloneFrom(this);
            return cloned;
        }

        public void CloneFrom(IModifier<int> source)
        {
            if (source is NumericIModifier other)
            {
                operation = other.operation;
                value = other.value;
            }
        }
    }
}

File: 유틸리티/데이터/Value/Lua/Modifier/NumericIModifier.lua
---@meta Value

---@enum NumericIOperation
local NumericIOperation = {
    SET = 0,
    ADD = 1,
    SUB = 2,
    MUL = 3,
    DIV = 4
}

---@class NumericIModifier : IModifier
--- # 프로퍼티
---@field Operation NumericIOperation @수치 연산 타입입니다.
---@field Value integer @연산에 사용될 정수형 값입니다.
--- # 메서드
---@field Modify fun(self, value: integer): integer @수치 연산을 적용합니다.
---@field CloneFrom fun(self, source: NumericIModifier) @다른 수정자로부터 값을 복제합니다.
